modules/dnn/src/onnx/onnx_graph_simplifier.cpp

```diff
@@ -1184,6 +1184,15 @@ void simplifySubgraphs(opencv_onnx::GraphProto& net)
     subgraphs.push_back(makePtr<NormalizeSubgraph5>());
 
     simplifySubgraphs(Ptr<ImportGraphWrapper>(new ONNXGraphWrapper(net)), subgraphs);
+
+    // remove training outputs for conformance tests
+    if (net.node().size() == 1) {
+        const auto &node = net.node(0);
+        const auto &op_type = node.op_type();
+        if (op_type == "LayerNormalization") {
+            net.mutable_output()->DeleteSubrange(1, 2);
+        }
+    }
```

Comment:
Isn't the problem in the test ONNX models? Shall they be exported in `eval()` mode?

Reply:
No. They are generated by the ONNX official. You can find most of them in https://github.com/onnx/onnx/tree/main/onnx/backend/test/data/node.

These models and tests are added via https://github.com/opencv/opencv/pull/21088.

Also related issue to check ONNX operator coverage: https://github.com/opencv/opencv/issues/21078.

Reply:
I checked all ONNX operators with optional outputs. Turns out there are only BatchNormalization, Dropout, LayerNormalization and MaxPool which have optional outputs for training. All except LayerNormalization have conformance models set for training mode and eval mode. Some of them have training outputs, some dont, so we can still use those without training outputs for conformance tests. But this is not the case with LayerNormalization. All LayerNormalization conformance models have training outputs. Blame ONNX for that.

Seems like this patch is meaningful only for `LayerNormalization`. What do you think?

Reply:
As tests passed and layers' outputs do not depend in train/eval phase on which they were exported to ONNX, I'm fine.

The only concern is that changes made in the unrelated graph fusion method. Can this be moved to `parseLayerNorm` from `onnx_importer.cpp`?

Reply:
> Can this be moved to parseLayerNorm from onnx_importer.cpp?

AFAIK, it is not feasible because for conformance tests like I present above their ouputs are attached to graph outputs, and we cannot modify graph proto inside `parseLayerNorm`.

---

> The only concern is that changes made in the unrelated graph fusion method.

I put conditions to ensure `node_size == 1`, `op_type == "LayerNormalization"` and `node_output_size > 1`. I think it should be safe for basically all models, since they have more than one node even after fusion.

Reply:
@fengyuentau, please try this:

```patch
--- a/modules/dnn/src/onnx/onnx_graph_simplifier.cpp
+++ b/modules/dnn/src/onnx/onnx_graph_simplifier.cpp
@@ -1184,18 +1184,6 @@ void simplifySubgraphs(opencv_onnx::GraphProto& net)
     subgraphs.push_back(makePtr<NormalizeSubgraph5>());
 
     simplifySubgraphs(Ptr<ImportGraphWrapper>(new ONNXGraphWrapper(net)), subgraphs);
-
-    // remove training outputs for conformance tests
-    if (net.node().size() == 1) {
-        const auto &node = net.node(0);
-        const auto &op_type = node.op_type();
-        if (op_type == "LayerNormalization") {
-            if (net.output_size() > 1) {
-                int num = net.output_size() - 1;
-                net.mutable_output()->DeleteSubrange(1, num);
-            }
-        }
-    }
 }
 
 Mat getMatFromTensor(const opencv_onnx::TensorProto& tensor_proto)
diff --git a/modules/dnn/src/onnx/onnx_importer.cpp b/modules/dnn/src/onnx/onnx_importer.cpp
index 28dd4d9b77..80f57c4156 100644
--- a/modules/dnn/src/onnx/onnx_importer.cpp
+++ b/modules/dnn/src/onnx/onnx_importer.cpp
@@ -3187,6 +3187,17 @@ void ONNXImporter::parseLayerNorm(LayerParams& layerParams, const opencv_onnx::N
     // Remove additional outputs (Mean, InvStdDev)
     if (node_proto.output_size() > 1)
     {
+        for (size_t i = 1; i < node_proto.output_size(); ++i)
+        {
+            for (int j = graph_proto.output_size() - 1; j >= 0; --j)
+            {
+                if (graph_proto.output(j).name() == node_proto.output(i))
+                {
+                    graph_proto.mutable_output()->DeleteSubrange(j, 1);
+                    break;
+                }
+            }
+        }
         auto outputName = node_proto.output(0);
         opencv_onnx::NodeProto node_proto_ = node_proto;
         node_proto_.clear_output();
```

Reply:
I did not know graph_proto is available inside each `parseXXX`. Thanks, it works.